File: solutions/0138. Copy List with Random Pointer/0138.py
class Solution:
  def copyRandomList(self, head: 'Node') -> 'Node':
    if head is None:
      return None
    if head in self.map:
      return self.map[head]

    newNode = Node(head.val)
    self.map[head] = newNode
    newNode.next = self.copyRandomList(head.next)
    newNode.random = self.copyRandomList(head.random)
    return newNode

  map = {}

File: solutions/0138. Copy List with Random Pointer/0138.cpp
class Solution {
 public:
  Node* copyRandomList(Node* head) {
    if (!head)
      return nullptr;
    if (map.count(head))
      return map[head];

    Node* newNode = new Node(head->val);
    map[head] = newNode;
    newNode->next = copyRandomList(head->next);
    newNode->random = copyRandomList(head->random);
    return newNode;
  }

 private:
  unordered_map<Node*, Node*> map;
};
